src/parking_lot.rs

```diff
@@ -0,0 +1,149 @@
+#[cfg(feature = "loom")]
+pub use self::with_loom::*;
+
+#[cfg(not(feature = "loom"))]
+pub use parking_lot::*;
+
+#[cfg(feature = "loom")]
+mod with_loom {
+	use std::{
+		fmt, mem,
+		mem::swap,
+		ops::{Deref, DerefMut},
+	};
+
+	#[derive(Debug, Default)]
+	pub struct Mutex<T>(loom::sync::Mutex<T>);
+
+	impl<T> Mutex<T> {
+		pub fn new(val: T) -> Self {
+			Self(loom::sync::Mutex::new(val))
+		}
+
+		pub fn lock(&self) -> MutexGuard<'_, T> {
+			MutexGuard(self.0.lock().unwrap())
+		}
+	}
+
+	#[derive(Debug)]
+	pub struct MutexGuard<'a, T>(loom::sync::MutexGuard<'a, T>);
+
+	impl<'a, T> Deref for MutexGuard<'a, T> {
+		type Target = T;
+
+		fn deref(&self) -> &T {
+			self.0.deref()
+		}
+	}
+
+	impl<'a, T> DerefMut for MutexGuard<'a, T> {
+		fn deref_mut(&mut self) -> &mut T {
+			self.0.deref_mut()
+		}
+	}
+
+	impl<'a, T: fmt::Display> fmt::Display for MutexGuard<'a, T> {
+		fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+			self.0.fmt(f)
+		}
+	}
+
+	#[derive(Debug, Default)]
+	pub struct Condvar(loom::sync::Condvar);
+
+	impl Condvar {
+		pub fn new() -> Self {
+			Self(loom::sync::Condvar::new())
+		}
+
+		pub fn notify_one(&self) {
+			self.0.notify_one();
+		}
+
+		pub fn notify_all(&self) {
+			self.0.notify_all()
+		}
+
+		pub fn wait<T>(&self, mutex_guard: &mut MutexGuard<'_, T>) {
+			// loom API takes mutex by value. We do a hack to get the value from the reference and
+			// copy back it at the end
+			#[allow(invalid_value)]
+			let mut owned_gard = unsafe { mem::zeroed() };
+			swap(&mut mutex_guard.0, &mut owned_gard);
+			owned_gard = self.0.wait(owned_gard).unwrap();
+			mutex_guard.0 = owned_gard;
+		}
+	}
+
+	#[derive(Debug, Default)]
+	pub struct RwLock<T>(loom::sync::RwLock<T>);
+
+	impl<T> RwLock<T> {
+		pub fn new(val: T) -> Self {
+			Self(loom::sync::RwLock::new(val))
+		}
+
+		pub fn read(&self) -> RwLockReadGuard<'_, T> {
+			RwLockReadGuard(self.0.read().unwrap())
+		}
+
+		pub fn upgradable_read(&self) -> RwLockUpgradableReadGuard<'_, T> {
+			RwLockUpgradableReadGuard(self.0.write().unwrap())
+		}
+
+		pub fn write(&self) -> RwLockWriteGuard<'_, T> {
+			RwLockWriteGuard(self.0.write().unwrap())
+		}
+	}
+
+	#[derive(Debug)]
+	pub struct RwLockReadGuard<'a, T>(loom::sync::RwLockReadGuard<'a, T>);
+
+	impl<'a, T> Deref for RwLockReadGuard<'a, T> {
+		type Target = T;
+
+		fn deref(&self) -> &T {
+			self.0.deref()
+		}
+	}
+
+	#[derive(Debug)]
+	pub struct RwLockUpgradableReadGuard<'a, T>(loom::sync::RwLockWriteGuard<'a, T>);
```

Comment:
The problem if we use a read gard is to do properly the swap to a write gard without giving a chance to an other thread to acquire a write gard. I prefered to be convervative rather than lenient in order to avoid finding not-existing bugs.

Reply:
yes I understand better the issue, would be worth being documented I think.

Reply:
Done

Reply:
I was wondering, probably not a good idea, but would using a fork of parking_lot that replace it's atomic with loom atomic be a solution?

Reply:
Isn't parking_lot also using primitives like `futex` on Linux to implement is `Mutex`? I fear that just using loom's atomics would not be good enough and we would need a way to intrument these primitives too. It's maybe possible to add to loom instrumentation of "upgreadable reads" based on the current Loom runtime. So, I would maybe fork loom instead of parking_lot.

Reply:
Yes :+1: , there is one futex usage for parking thread, may be doable to use loom park at this level or just plainly wrong. Plugging things at atomic level would make things closer to the actual behavior (probably would need to skip timer too), but anyway does not seems to me like a maintainable fork.